scripts/zowe-install-extension.sh

```diff
@@ -0,0 +1,148 @@
+
+install_desktop_plugin(){
+    cd $TARGET_DIR
+
+    if [ ! -d "$EXTENSION_NAME" ]; then
+        mkdir $EXTENSION_NAME
+    else
+        # Clean up the directory and make a new one
+        rm -rf $EXTENSION_NAME
+        mkdir $EXTENSION_NAME
+    fi
+
+    cd $EXTENSION_NAME
+
+    pax -ppx -rf "$EXTENSION_PATH"
+
+    # Validates the pax file is correctly configured when unpaxed
+    if [ ! -e "pluginDefinition.json" ]; then
+        error_handler "$EXTENSION_PATH: Invalid package"
+    fi
+
+    # Uses install-app.sh in zowe-instance-dir to automatically set up the extension onto zowe
+    $INSTANCE_DIR/bin/install-app.sh $TARGET_DIR/$EXTENSION_NAME
+
+    echo "Zowe desktop plugin has been installed."
+}
+
+while [ $# -gt 0 ]; do #Checks for parameters
+    arg="$1"
+    case $arg in
+        -e|--extension) #Represents the path pointed to the extension's pax file
+        shift
+        if [[ "$1" = *.pax ]]; then
+            EXTENSION_PATH=$1
+        else
+            error_handler "-e|--extension: Given path is not a pax file or does not exist"
+        fi
+        shift
+        ;;
+        -i|--instance_dir) #Represents the path to zowe's instance directory 
+        shift
+        if [ -d "$1" ] && [ -e "$1/instance.env" ]; then
+            INSTANCE_DIR=$1
+        else
+            error_handler "-i|--instance_dir: Given path is not a zowe instance directory or does not exist"
+        fi
+        shift
+        ;;
+        -d|--target_dir)
+        shift
+        TARGET_DIR=$1
+        shift
+        ;;
+        -n|--name) #Represents the name of the plugin/extension
+        shift
+        EXTENSION_NAME="$1"
+        shift
+        ;;
+        -t|--type) #Represents the type of extension whether its a desktop plugin or web component
+        shift
+        if [[ $1 == "component" ]] || [[ $1 == "desktop-plugin" ]]; then
+            INSTALL_TYPE=$1
+        else
+            error_handler "-t|--type: Invalid value, possible values are 'component' and 'desktop-plugin'"
+        fi
+        shift
+        ;;
+        *)
+        error_handler "$1 is an invalid flag\ntry: zowe-install-extensions.sh -e {PATH_TO_EXTENSION} -i {ZOWE_INSTANCE_DIR} -n {NAME_OF_EXTENSION} -t component|desktop-plugin"
+    esac
+done
+
+if [ -z $EXTENSION_PATH ] || [ -z $INSTANCE_DIR ] || [ -z $EXTENSION_NAME ] || [ -z $INSTALL_TYPE ]; then
+    #Ensures that the required parameters are entered, otherwise exit the program
+    error_handler "Missing parameters, try: zowe-install-extensions.sh -e {PATH_TO_EXTENSION} -i {ZOWE_INSTANCE_DIR} -n {NAME_OF_EXTENSION} -t component|desktop-plugin"
+fi
+
+if [ -z $TARGET_DIR ]; then
+    #Assigns TARGET_DIR to the default directory since it was not set to a specific directory
+    TARGET_DIR="$INSTANCE_DIR/extensions"
+fi
+
+if [ ! -e $TARGET_DIR ]; then
+    echo "Creating extensions folder at $TARGET_DIR"
+    mkdir $TARGET_DIR
+fi
+
+if [[ $INSTALL_TYPE == "component" ]]; then
+    install_component
+else
```

Comment:
What about APIML static definitions?
What about iframe apps?

jes-explorer is both an app, a component, and registers a static definition and creates an iframe object. Should these tasks be moved out of the jes-explorer code and into this script?

Reply:
APIML definitions and IFrame apps won't make differences in `install` and `configureInstance` stages. Those information can also be defined in manifest, but will be handled at `configure` stage, which is part of runtime, because they requires valid value of `ZOWE_EXPLORER_HOST` and `STATIC_DEF_CONFIG_DIR`.